source/agora/node/FullNode.d

```diff
@@ -466,4 +495,65 @@ public class FullNode : API
     protected void onAcceptedTransaction () @safe
     {
     }
+
+    /***************************************************************************
+
+        Push the block data to the `block_handlers` target server list
+        set in config.
+
+        Convert block data to JSON serialization and send it POST using Rest.
+
+        Params:
+            block = externalized block data
+
+    ***************************************************************************/
+
+    private void pushBlock (const Block block) @safe
+    {
+        foreach (address, handler; this.block_handlers)
+        {
+            runTask({
+                try
+                {
+                    handler.pushBlock(block);
+                }
+                catch (Exception e)
+                {
+                    log.error("Error sending block height #{} to {} :{}",
+                        block.header.height, address, e);
+                }
+            });
+        }
+    }
+
+    /***************************************************************************
+
+        Push the preimage to the `preimage_handlers` target server list
+        set in config.
+
+        Convert `PreImageInfo` to JSON serialization and send it POST
+        using Rest.
+
+        Params:
+            preImage = Received `PreImageInfo`
+
+    ***************************************************************************/
+
+    private void pushPreImage (const PreImageInfo pre_image) @safe
+    {
+        foreach (address, handler; this.preimage_handlers)
+        {
+            runTask({
+                try
+                {
+                    handler.pushPreImage(pre_image);
```

Comment:
I have a problem here.
When matching 'PreImageInfo' information with 'enrolled enrolment' in 'Stoa'
If there are multiple `enrolled enrolments` with the same UTXO_key,
We have to think about which 'PreImageCycle' matches which of the two other 'Enrollments'.
Because there is no information about 'PreImageCycle' in 'Stoa'.
Therefore, we need to make sure that 'Hashing' matches 'Random seed' of 'Enrollment' by doing 'Distance' of 'PreImageInfo'.
However, it can be inefficient if Stoa does Hashing every time he receives 'Preimage'.

Reply:
> If there are multiple enrolled enrolments with the same UTXO_key,

Multiple enrollments with the same UTXO? That shouldn't be possible.

Reply:
He means continuous enrollments with the same UTXO. He's worried about the situation where some application could proceed during a period of more than one enrollment. So he wants to know what's the enrollment for a specific pre-image.

Reply:
> Multiple enrollments with the same UTXO? That shouldn't be possible.

@AndrejMitrovic  Yes, it's possible. Most likely to use the same UTXO.

@linked0  That's right.

Reply:
For example, multiple enrollments mean to the PreImageCycle 0 enrollment and the PreImageCycle 1 enrollment.
Here's what Jay describes as a continuous enrollments.

Reply:
The word `multiple enrollments` seems to cause confusion, so I'll call it `continuous  enrollments`

Reply:
But they cannot be active at the same block height, right?

Reply:
> But they cannot be active at the same block height, right?

Yes, I'd like to believe that way.

Reply:
So is there a problem here or no? Is the PR ready to go?

Reply:
> So is there a problem here or not? Is the PR ready to go?

I think Stoa needs to figure it out.